From: Jan Kara <jack@suse.cz>
Subject: proc: Show DAX mappings as having VM_MIXEDMAP flag
Patch-mainline: Never, hack for HANA compatibility
References: bsc#1110360

Until commit e1fb4a08649 "dax: remove VM_MIXEDMAP for fsdax and device dax",
userspace could distinguish DAX and non-DAX mappings in /proc/<pid>/smaps by
checking whether the VMA has VM_MIXEDMAP (mm in smaps) flag. Apparently SAP
HANA uses this to verify it indeed got a DAX mapping. Fake the 'mm' flag for it
for now.

Signed-off-by: Jan Kara <jack@suse.cz>

---
 fs/proc/task_mmu.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

--- a/fs/proc/task_mmu.c
+++ b/fs/proc/task_mmu.c
@@ -18,6 +18,7 @@
 #include <linux/shmem_fs.h>
 #include <linux/uaccess.h>
 #include <linux/pkeys.h>
+#include <linux/fs.h>
 
 #include <asm/elf.h>
 #include <asm/tlb.h>
@@ -684,7 +685,8 @@ static void show_smap_vma_flags(struct s
 	for (i = 0; i < BITS_PER_LONG; i++) {
 		if (!mnemonics[i][0])
 			continue;
-		if (vma->vm_flags & (1UL << i)) {
+		if ((vma->vm_flags & (1UL << i)) ||
+		    ((1UL << i) == VM_MIXEDMAP && vma_is_dax(vma))) {
 			seq_printf(m, "%c%c ",
 				   mnemonics[i][0], mnemonics[i][1]);
 		}
